cmd/internal/labeler/labeler.go

```diff
@@ -455,8 +471,97 @@ func createNumaNodeMappingLabel(mapping map[int][]string) string {
 	return strings.Join(parts, "_")
 }
 
-func (l *labeler) printLabels() {
+func (l *labeler) printLabelsToFile(labelFile string) error {
+	f, err := os.OpenFile(labelFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		klog.Warning("Failed to open file:", err)
+
+		return err
+	}
+
+	for key, val := range l.labels {
+		if _, err := f.WriteString(key + "=" + val + "\n"); err != nil {
+			klog.Warning("Failed to write label to file:", err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+func CreateAndPrintLabels(sysfsDRMDir, debugfsDRIDir string) {
+	l := newLabeler(sysfsDRMDir, debugfsDRIDir)
+
+	if err := l.createLabels(); err != nil {
+		klog.Warningf("failed to create labels: %+v", err)
+
+		return
+	}
+
 	for key, val := range l.labels {
 		fmt.Println(key + "=" + val)
 	}
 }
+
+// Gathers node's GPU labels on channel trigger or timeout, and write them to a file.
+// The created label file is deleted on exit (process dying).
+func Run(pathPrefix string, updateInterval time.Duration, scanResources chan bool) {
+	l := newLabeler(pathPrefix+sysfsDRMDirectory, pathPrefix+debugfsDRIDirectory)
+
+	interruptChan := make(chan os.Signal, 1)
+	signal.Notify(interruptChan, os.Interrupt)
+	signal.Notify(interruptChan, syscall.SIGTERM)
+
+	klog.V(1).Info("Starting GPU labeler")
+
+	running := true
+
+	for running {
+		timeout := time.After(updateInterval)
+
+		select {
+		case <-timeout:
+		case <-scanResources:
+		case <-interruptChan:
+			klog.V(2).Info("Interrupt received")
+
+			running = false
+
+			continue
+		}
+
+		klog.V(1).Info("Ext resources scanning")
+
+		err := l.createLabels()
+		if err != nil {
+			klog.Warningf("label creation failed: %+v", err)
+
+			continue
+		}
+
+		if l.labelsChanged {
+			klog.V(1).Info("Writing labels")
+
+			if err := l.printLabelsToFile(nfdFeatureFile); err != nil {
+				klog.Warningf("failed to write labels to file: %+v", err)
+			}
+		}
+	}
+
+	signal.Stop(interruptChan)
+
+	klog.V(2).Info("Removing label file")
+
+	err := os.Remove(nfdFeatureFile)
+	if err != nil {
+		klog.Errorf("Failed to cleanup label file: %+v", err)
+	}
+
+	klog.V(1).Info("Stopping GPU labeler")
+
+	// Trigger sigterm to cause the app to exit without delays
+	if err := syscall.Kill(syscall.Getpid(), syscall.SIGTERM); err != nil {
```

Comment:
Why not just `os.Exit(0)`?

Reply:
I didn't want to exit the whole application in case there is some other signal handling happening elsewhere in the code.

Reply:
Oh, this a go-thread?   Then you should be using a `channel` instead, and listen on that where you listen for external signals (main thread?).

Signal delivery is asynchronous method _separate_ from threading, so locks don't work in signal handlers, and you can use only very few syscalls & functions safely in them (see "man 7 signal").  Golang probably uses something like pipes to do signal notifications from signal handler to main thread.  IMHO better just to use language feature designed for such notifications directly.

Reply:
I don't think this context is within the signal handling. The signal itself is delivered via channel, so I'd imagine the limitations about signal handling doesn't apply anymore.

In any case, I changed the labeler to call `Exit()` directly and I also added comments about the flow to the gpu plugin's main. I couldn't find any signal handling in the rest of the device plugin code. And the device plugin manager would need refactoring to support clean exit (from a channel).